File: ubuntu/catkin_ws/src/bionic_hand/src/move_motor.cpp
#include <iostream>
#include <dynamixel_sdk/dynamixel_sdk.h>
#include <ros/ros.h>
#include <thread>
#include <bionic_hand/ControlCommands.h>  //ROS message header
#include <unistd.h>

// Dynamixel settings
#define PROTOCOL_VERSION 2.0
#define DXL_ID 1
#define BAUDRATE 57600
#define DEVICENAME "/dev/ttyUSB1"
#define ADDR_PRO_GOAL_PWM 100
#define ADDR_PRO_TORQUE_ENABLE 64
#define ADDR_PRO_GOAL_POSITION 116
#define ADDR_PRESENT_POSITION 132
#define ADDR_OPERATING_MODE 11
#define TORQUE_ENABLE 1
#define TORQUE_DISABLE 0
#define MAX_PWM 885
#define POSITION_MODE 3
#define PWM_MODE 16

// Global variable for PWM value
float g_PWM = 0;

// Initialize Dynamixel
dynamixel::PortHandler* initializeDynamixel(const char* devicename, int baudrate) {
    auto portHandler = dynamixel::PortHandler::getPortHandler(devicename);
    if (!portHandler->openPort()) {
        throw std::runtime_error("Failed to open the port");
    }
    if (!portHandler->setBaudRate(baudrate)) {
        throw std::runtime_error("Failed to change the baudrate");
    }
    return portHandler;
}

// ROS callback function
void controlCommandCallback(const bionic_hand::ControlCommands::ConstPtr& msg) {
    g_PWM = msg->PWM;
}

float read_position(dynamixel::PacketHandler* packetHandler, dynamixel::PortHandler* portHandler, int dxl_ID, int addr_presnt_position){
    uint32_t dxl_present_position;
    uint8_t dxl_error = 0;
    int dxl_comm_result = packetHandler->read4ByteTxRx(portHandler, dxl_ID, addr_presnt_position, &dxl_present_position, &dxl_error);
    if (dxl_comm_result != COMM_SUCCESS) {
        packetHandler->getTxRxResult(dxl_comm_result);
        return -1.0; //indicate an error
    } else if (dxl_error != 0) {
        packetHandler->getRxPacketError(dxl_error);
        return -1.0; //indicate an error
    }
    // Convert the position reading if necessary (depends on Dynamixel model and firmware)
    if (dxl_present_position > 2147483647){  // 2147483647 is 2^31 - 1, half of the range of 32-bit signed integer
        dxl_present_position -= 4294967296;  // Subtract 2^32 to get the correct negative value
    }
    //  Convert position from raw value to degrees if necessary (depends on Dynamixel model)
    dxl_present_position = dxl_present_position * 0.088;  // Example conversion factor to degrees (may vary)
    
    
    return dxl_present_position; //return current position
}

void set_goal_position(dynamixel::PacketHandler* packetHandler, dynamixel::PortHandler* portHandler, int dxl_id,  float position){
    int dxl_comm_result; // Result of communication
    uint8_t dxl_error = 0; // Dynamixel error

    // Assuming goal position is a 4-byte value; change to write2ByteTxRx if it's a 2-byte value
    dxl_comm_result = packetHandler->write4ByteTxRx(portHandler, dxl_id, ADDR_PRO_GOAL_POSITION, position, &dxl_error);

    if (dxl_comm_result != COMM_SUCCESS) {
        std::cerr << "Failed to set goal position: " << packetHandler->getTxRxResult(dxl_comm_result) << std::endl;
        throw std::runtime_error("Failed to set goal position");
    }
    if (dxl_error != 0) {
        std::cerr << "Dynamixel error: " << packetHandler->getRxPacketError(dxl_error) << std::endl;
    } else {
        std::cout << "Moving robot to starting position: " << position << std::endl;
    }
}

// Main function
int main(int argc, char **argv) {
    std::cout << "Prepared to move motor" << std::endl;

    
    ros::init(argc, argv, "move_motor_node");
    ros::NodeHandle nh;
    ros::Subscriber sub = nh.subscribe("Control_Command", 1000, controlCommandCallback);

    auto portHandler = initializeDynamixel(DEVICENAME, BAUDRATE);
    auto packetHandler = dynamixel::PacketHandler::getPacketHandler(PROTOCOL_VERSION);

    // # Set operating mode to Position Control Mode
    packetHandler->write1ByteTxRx(portHandler, DXL_ID, ADDR_OPERATING_MODE, POSITION_MODE);
    packetHandler->write1ByteTxRx(portHandler, DXL_ID, ADDR_PRO_TORQUE_ENABLE, TORQUE_ENABLE);
    float starting_point = 180;
    float final_point = 350;
    float goal_position = int(starting_point / final_point * 4095);  // Convert 180 degrees to the corresponding value (assuming 12-bit resolution)
    std::cout<<"Moving robot to starting position: " << goal_position<< std::endl;
    set_goal_position(packetHandler, portHandler, DXL_ID, goal_position);
    sleep(1);
    packetHandler->write1ByteTxRx(portHandler, DXL_ID, ADDR_PRO_TORQUE_ENABLE, TORQUE_DISABLE); //disable torequ
    std::cout<<"Begin Control "<< std::endl;

    // Set operating mode to PWM
    packetHandler->write1ByteTxRx(portHandler, DXL_ID, ADDR_OPERATING_MODE, PWM_MODE);

    // Enable torque
    packetHandler->write1ByteTxRx(portHandler, DXL_ID, ADDR_PRO_TORQUE_ENABLE, TORQUE_ENABLE);

    // Main loop
    while (ros::ok()) {
        // std::cout << "driving motor" << std::endl;
        
        int dxl_present_position = read_position(packetHandler, portHandler, DXL_ID, ADDR_PRESENT_POSITION);

        // std::cout << "Current Position: " << dxl_present_position << " -- PWM: " << g_PWM << std::endl;
        if (dxl_present_position < 175 || dxl_present_position > 340) {
            packetHandler->write2ByteTxRx(portHandler, DXL_ID, ADDR_PRO_GOAL_PWM, 0);
            std::cout<<"Motor limit reached. Stopping at: "<< dxl_present_position<< std::endl;
            break;

        }
        else{
            packetHandler->write2ByteTxRx(portHandler, DXL_ID, ADDR_PRO_GOAL_PWM, g_PWM);
        }
        // ros::Rate rate(1.0 / 0.005); // rate = 10 Hz if dt = 0.1

        ros::spinOnce();
    }

    // Disable torque and close port
    packetHandler->write1ByteTxRx(portHandler, DXL_ID, ADDR_PRO_TORQUE_ENABLE, TORQUE_DISABLE);
    portHandler->closePort();

    return 0;
}

File: ubuntu/catkin_ws/src/bionic_hand/src/move_motor.py
import serial
import csv
import os
import dynamixel_sdk as dxl
import time
import queue
# from scipy import signal
import numpy as np
from threading import Thread
import rospy

# Protocol version and Dynamixel settings
PROTOCOL_VERSION = 2.0
DXL_ID = 1
BAUDRATE = 57600
DEVICENAME = '/dev/ttyUSB0'
ADDR_PRO_GOAL_PWM = 100
ADDR_PRO_TORQUE_ENABLE = 64
ADDR_PRO_GOAL_POSITION = 116
ADDR_PRESENT_POSITION = 132
ADDR_OPERATING_MODE = 11  # Address for the operating mode
TORQUE_ENABLE = 1
TORQUE_DISABLE = 0
MAX_VOLTAGE = 12.2
MAX_PWM = 885
step_input = 12  # input voltage
POSITION_MODE = 3
PWM_MODE = 16
step_magnitude = 5 #volts
from bionic_hand.msg import ControlCommands


# Function to initialize Dynamixel
def initialize_dynamixel(devicename, baudrate, protocol_version):
    portHandler = dxl.PortHandler(devicename)
    packetHandler = dxl.PacketHandler(protocol_version)
    if not portHandler.openPort():
        raise IOError("Failed to open the port")
    if not portHandler.setBaudRate(baudrate):
        raise IOError("Failed to change the baudrate")
    return portHandler, packetHandler

# Function to set operating mode
def set_operating_mode(packet_handler, port_handler, mode):
    disable_torque(packet_handler, port_handler, DXL_ID)
    dxl_comm_result, dxl_error = packet_handler.write1ByteTxRx(port_handler, DXL_ID, ADDR_OPERATING_MODE, mode)
    if dxl_comm_result != dxl.COMM_SUCCESS:
        raise Exception("Failed to set operating mode")
    if dxl_error != 0:
        print("Dynamixel error: %s" % packet_handler.getRxPacketError(dxl_error))


# Function to enable torque
def enable_torque(packetHandler, portHandler, dxl_id, torque_enable):
    dxl_comm_result, dxl_error = packetHandler.write1ByteTxRx(portHandler, dxl_id, ADDR_PRO_TORQUE_ENABLE, torque_enable)
    if dxl_comm_result != dxl.COMM_SUCCESS:
        raise RuntimeError(f"Torque enable failed: {packetHandler.getTxRxResult(dxl_comm_result)}")

# Function to disable torque
def disable_torque(packetHandler, portHandler, dxl_id):
    packetHandler.write1ByteTxRx(portHandler, dxl_id, ADDR_PRO_TORQUE_ENABLE, TORQUE_DISABLE)


# Function to close port
def close_port(portHandler):
    portHandler.closePort()

def read_position(packet_handler, port_handler, dxl_id, addr_present_position):
    dxl_present_position, dxl_comm_result, dxl_error = packet_handler.read4ByteTxRx(port_handler, dxl_id, addr_present_position)
    if dxl_comm_result != dxl.COMM_SUCCESS or dxl_error != 0:
        print("Failed to read position")
        return None
    
    # Convert the position reading if necessary (depends on Dynamixel model and firmware)
    if dxl_present_position > 2147483647:  # 2147483647 is 2^31 - 1, half of the range of 32-bit signed integer
        dxl_present_position -= 4294967296  # Subtract 2^32 to get the correct negative value

    # Convert position from raw value to degrees if necessary (depends on Dynamixel model)
    dxl_present_position = dxl_present_position * 0.088  # Example conversion factor to degrees (may vary)
    
    return dxl_present_position



# Function to set goal position
def set_goal_position(packet_handler, port_handler, position):
    dxl_comm_result, dxl_error = packet_handler.write4ByteTxRx(port_handler, DXL_ID, ADDR_PRO_GOAL_POSITION, position)
    if dxl_comm_result != dxl.COMM_SUCCESS:
        raise Exception("Failed to set goal position")
    if dxl_error != 0:
        print("Dynamixel error: %s" % packet_handler.getRxPacketError(dxl_error))


# Function to set PWM
def set_pwm(packet_handler, port_handler, dxl_id, addr_pro_goal_pwm, pwm):
    dxl_comm_result, dxl_error = packet_handler.write2ByteTxRx(port_handler, dxl_id, addr_pro_goal_pwm, pwm)
    if dxl_comm_result != dxl.COMM_SUCCESS or dxl_error != 0:
        print("Failed to set PWM: %s, %s" % (packet_handler.getTxRxResult(dxl_comm_result), packet_handler.getRxPacketError(dxl_error)))
        return False
    print(f"Set PWM to {pwm}")
    return True


class FingerControlCommandSubscriber:
    def __init__(self):
        # Initialize the ROS node
        # rospy.init_node('control_command_subscriber', anonymous=True)

        # Variable to hold the position

        self.PWM = 0

        # Set up the subscriber
        # The callback function is called every time a message is received
        self.subscriber = rospy.Subscriber("Control_Command", ControlCommands, self.update_control_command)
        self.thread = Thread(target=self.spin) #create a thread for the spin method so it runs independantly of the rest of the code
        self.thread.daemon = True #thread will exit when main program is terminated
        self.thread.start() # start thread

    def update_control_command(self, msg):
        """Callback function to handle incoming messages."""
        # self.finger_position = msg
        self.PWM = msg.PWM

        # rospy.loginfo(f"Updated finger position: {msg}")

    def spin(self):
        """Keep the node running so it can keep receiving messages."""
        rospy.spin()
    
    def get_latest_control_command(self):
        # print(self.PWM)

        return self.PWM
    

def main():
    rospy.init_node('move_motor_node', anonymous=True)
    fccs = FingerControlCommandSubscriber()

    portHandler, packetHandler = None, None
    print("Motor 1 ready to move")

    try:

        # Initialize PortHandler and PacketHandler
        portHandler, packetHandler = initialize_dynamixel(DEVICENAME, BAUDRATE, PROTOCOL_VERSION)

        # Switch back to PWM mode
        set_operating_mode(packetHandler, portHandler, PWM_MODE)
        enable_torque(packetHandler, portHandler, DXL_ID, TORQUE_ENABLE)
        pwm = int(fccs.get_latest_control_command())
        # print("pwm: ", pwm)

        # Main loop to set PWM values
        while True:
            dxl_present_position = read_position(packetHandler, portHandler, DXL_ID, ADDR_PRESENT_POSITION)
            if dxl_present_position < 175 or dxl_present_position > 250: #if the motor reaches max safe pos stop. CREATE A SEPERATE FILE/FUNCTION FOR THIS
                print("Stopping pos: ", dxl_present_position)
                set_pwm(packetHandler, portHandler, DXL_ID, ADDR_PRO_GOAL_PWM, pwm = 0) #stop 
                break
            if not set_pwm(packetHandler, portHandler, DXL_ID, ADDR_PRO_GOAL_PWM, pwm): #set pwm to whatever is on the pwms list
                continue
            # time.sleep(dt)





    except Exception as e:
        print(e)

    finally:
        if packetHandler and portHandler:
            disable_torque(packetHandler, portHandler, DXL_ID)
            close_port(portHandler)

if __name__ == "__main__":
    main()
